### ISSUE TITLE
Resilience and Migration

### ISSUE THREAD

**Issue by alrossi** [dCache/dcache#5637]
I am opening the issue more as a matter of reference, though there is one small non-fatal flaw in the way resilience reacts when migrating a resilient pool.

The following is a point-by-point description of a test conducted on a small test stand involving a simple migration of files from a source pool to a target pool, both of which are part of a resilient pool group.

**1) First, we check which files on the source pool are already on the target pool.**

```
\s Resilience contained in dcatest04-4|dcatest08-3
Wed Dec 16 08:08:08 CST 2020         290fc3b7-e7d3-4c3b-9338-d41bbcc7a392    contained-in scan, file name: contained-in-1608127688464 (RUNNING)

\s Resilience async cmd ls
Wed Dec 16 08:08:08 CST 2020         290fc3b7-e7d3-4c3b-9338-d41bbcc7a392    contained-in scan, file name: contained-in-1608127688464 (DONE)

[root@fndcatemp2 ~]# cd /var/lib/dcache/resilience/
[root@fndcatemp2 resilience]# sort contained-in-1608127688464 | uniq > shared-04-4-08-3.txt
[root@fndcatemp2 resilience]# wc -l shared-04-4-08-3.txt
123 shared-04-4-08-3.txt
```

**2) Next, let us record the current contents of the source and target pool repositories.**

```
echo "\s dcatest04-4 rep ls" | ssh -p 24223 admin@fndcatemp2.fnal.gov | cut -d ' ' -f1 > dcatest04-4-before

Alberts-MacBook-Pro:drain arossi$ wc -l dcatest04-4-before
    1127 dcatest04-4-before

echo "\s dcatest08-3 rep ls" | ssh -p 24223 admin@fndcatemp2.fnal.gov | cut -d ' ' -f1 > dcatest08-3-before

Alberts-MacBook-Pro:drain arossi$ wc -l dcatest08-3-before 
    1126 dcatest08-3-before
```

**3) We will now initiate the drain of the source pool by setting it to read only and starting a migration move job.**

```
\s dcatest04-4 pool disable -rdonly
\s dcatest04-4 migration move -target=pool dcatest08-3
\s dcatest04-4 migration ls
[1] FINISHED     migration move -target=pool -- dcatest08\-3
```

**4) Next, record the contents of the target pool.**

```
echo "\s dcatest08-3 rep ls" | ssh -p 24223 admin@fndcatemp2.fnal.gov | cut -d ' ' -f1 > dcatest08-3-after

Alberts-MacBook-Pro:drain arossi$ wc -l dcatest08-3-after
    2130 dcatest08-3-after
```

2130=1127+1126-123, so purely from the standpoint of arithmetic, it checks out: the files that were on dcatest04-4 are now all on dcatest08-3.

**5) But let's look at what resilience actually did.**  

The transfer was done on a freshly started resilience domain, so history was empty before the transfer was started.

```
echo "\s Resilience history" | ssh -p 24223 admin@fndcatemp2.fnal.gov > resilience-history.txt
```

Parsing out the pnfsids for copy, remove and sticky operations.

```
Alberts-MacBook-Pro:Desktop arossi$ grep COPY resilience-history.txt | cut -d ' ' -f3 | cut -c 2- | sort > copy.txt
Alberts-MacBook-Pro:drain arossi$ wc -l copy.txt 
      235 copy.txt

Alberts-MacBook-Pro:Desktop arossi$ grep REMOVE resilience-history.txt | cut -d ' ' -f3 | cut -c 2- | sort > remove.txt
Alberts-MacBook-Pro:Desktop arossi$ wc -l remove.txt 
       6 remove.txt

Alberts-MacBook-Pro:Desktop arossi$ grep STICKY resilience-history.txt | cut -d ' ' -f3 | cut -c 2- | sort > sticky.txt
Alberts-MacBook-Pro:Desktop arossi$ wc -l sticky.txt 
       1 sticky.txt

Alberts-MacBook-Pro:drain arossi$ grep NOOP resilience-history.txt | cut -d ' ' -f3 | cut -c 2- | sort > noop.txt
Alberts-MacBook-Pro:drain arossi$ wc -l noop.txt 
     882 noop.txt
```

It did 235 copies, 6 removes (caches), and 1 reset to sticky.

Now those counts all represent the "final state" of an operations.   That is, if an operation does copy, then remove, it appears in the remove list, but not the copy list. 

The only overlap is with removes and sticky.

```
Alberts-MacBook-Pro:drain arossi$ for f in `cat sticky.txt`; do line=`grep $f remove.txt`; if [ "$line" == '' ] ; then echo "$f N"; else echo "$f Y"; fi; done | grep Y | wc -l
       1
Alberts-MacBook-Pro:drain arossi$ for f in `cat remove.txt`; do line=`grep $f copy.txt`; if [ "$line" == '' ] ; then echo "$f N"; else echo "$f Y"; fi; done | grep Y | wc -l
       0
Alberts-MacBook-Pro:drain arossi$ for f in `cat remove.txt`; do line=`grep $f noop.txt`; if [ "$line" == '' ] ; then echo "$f N"; else echo "$f Y"; fi; done | grep Y | wc -l
       0
Alberts-MacBook-Pro:drain arossi$ for f in `cat copy.txt`; do line=`grep $f noop.txt`; if [ "$line" == '' ] ; then echo "$f N"; else echo "$f Y"; fi; done | grep Y | wc -l
       0
```

**6) Before we address the removes and set sticky operations, let's take a look at the general status of the migrated files.**

First, let us check that all the files that have been drained from the source actually have more than one replica.

```
Alberts-MacBook-Pro:drain arossi$  for f in `cat dcatest04-4-before`; do echo $f >> namespace-check.txt; echo "\sn cacheinfoof $f" | ssh -p 24223 admin@fndcatemp2.fnal.gov >> namespace-check.txt; done

Alberts-MacBook-Pro:drain arossi$ grep dcatest08-3 namespace-check.txt | wc -l
    1127

Alberts-MacBook-Pro:drain arossi$ grep dcatest08-3 namespace-check.txt > locations.txt

Alberts-MacBook-Pro:drain arossi$ while IFS= read -r line; do echo $line | wc -w; done < locations.txt >> location-counts.txt

Alberts-MacBook-Pro:drain arossi$ grep 2 location-counts.txt | wc -l
     978
Alberts-MacBook-Pro:drain arossi$ grep 3 location-counts.txt | wc -l
     149
Alberts-MacBook-Pro:drain arossi$ grep 1 location-counts.txt | wc -l
       0
```

So all the files on the source have at least 2 locations after the migration.

Now, let us inspect the sticky bit status of the files.  Before resilience ran, the pools were all purged of cached files, so those detected now are the product of migration.  Are there at least two sticky per pnfsid?

```
Alberts-MacBook-Pro:drain arossi$ echo "\s dcatest* rep ls" | ssh -p 24223 admin@fndcatemp2.fnal.gov | grep -v ':' | grep -v X | cut -d ' ' -f5 | sort > not-sticky.txt 

Alberts-MacBook-Pro:drain arossi$ wc -l not-sticky.txt 
     360 not-sticky.txt

Alberts-MacBook-Pro:drain arossi$ for f in `cat not-sticky.txt`; do echo $f >> not-sticky-check.txt; echo "\sn cacheinfoof $f" | ssh -p 24223 admin@fndcatemp2.fnal.gov >> not-sticky-check.txt; done

Alberts-MacBook-Pro:drain arossi$ grep dcatest not-sticky-check.txt | wc -l
     360
```

This is the number of files with a cached replica.  The number of those that have three locations:

```
Alberts-MacBook-Pro:drain arossi$ grep dcatest not-sticky-check.txt | cut -d ' ' -f 4 | grep dcatest | wc -l
     355
```

This means that there are five files with only two locations, one of which must be cached.  That is, five files which did not get their final replica:

```
00001EFBA2E0EF324C358FAE701C645DA915
 dcatest03-4 dcatest08-3

\s dcatest03-4,dcatest08-3 rep ls 00001EFBA2E0EF324C358FAE701C645DA915
dcatest03-4:
    00001EFBA2E0EF324C358FAE701C645DA915 <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest08-3:
    00001EFBA2E0EF324C358FAE701C645DA915 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}

00003DC586211240464FA19B322FE51DE695
 dcatest07-3 dcatest08-3

[fndcatemp2] (Resilience@resilienceDomain) admin > \s dcatest07-3,dcatest08-3 rep ls 00003DC586211240464FA19B322FE51DE695
dcatest07-3:
    00003DC586211240464FA19B322FE51DE695 <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest08-3:
    00003DC586211240464FA19B322FE51DE695 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}

0000B65CB3979DC648CDBCF40D70AF69563E
 dcatest08-3 dcatest09-3

[fndcatemp2] (Resilience@resilienceDomain) admin > \s dcatest08-3,dcatest09-3 rep ls 0000B65CB3979DC648CDBCF40D70AF69563E
dcatest08-3:
    0000B65CB3979DC648CDBCF40D70AF69563E <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest09-3:
    0000B65CB3979DC648CDBCF40D70AF69563E <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}

0000C886B64D90ED44F1B6310980BEE67DC0
 dcatest08-3 dcatest09-4

[fndcatemp2] (Resilience@resilienceDomain) admin > \s dcatest08-3,dcatest09-4 rep ls 0000C886B64D90ED44F1B6310980BEE67DC0
dcatest08-3:
    0000C886B64D90ED44F1B6310980BEE67DC0 <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest09-4:
    0000C886B64D90ED44F1B6310980BEE67DC0 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}

0000E8FDA1F8455A460991C09FFB6BA43F3C
 dcatest08-3 dcatest09-4

[fndcatemp2] (Resilience@resilienceDomain) admin > \s dcatest08-3,dcatest09-4 rep ls 0000E8FDA1F8455A460991C09FFB6BA43F3C
dcatest08-3:
    0000E8FDA1F8455A460991C09FFB6BA43F3C <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest09-4:
    0000E8FDA1F8455A460991C09FFB6BA43F3C <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}
```

As it turns out, these are all files for which a "remove" operation was recorded.

```
Alberts-MacBook-Pro:drain arossi$ cat remove.txt 
00001EFBA2E0EF324C358FAE701C645DA915
00003DC586211240464FA19B322FE51DE695
0000B65CB3979DC648CDBCF40D70AF69563E
0000C886B64D90ED44F1B6310980BEE67DC0
0000E8FDA1F8455A460991C09FFB6BA43F3C
**0000F4088249698F49C69B907DDE40907C4B**
```

The last removed file is in fact the file which was also given a sticky bit operation:

```
Alberts-MacBook-Pro:drain arossi$ cat sticky.txt 
0000F4088249698F49C69B907DDE40907C4B
```

How do we explain this?

The removes are putatively the result of the following sequence of actions:

- AddCacheLocation message received as result of migration move.
- Resilience sees extra replica, and so caches one.
- Somehow the Clear Cache Location message received from migration move which actually deletes the source pool replica arrives just after the Resilience remove operation completes and terminates, thus neither being seen on the last namespace verification, nor triggering a second operation.

The sixth file, where a replica is set back to sticky, represents the latter case,
where the remove finishes, and then the clear cache location message arrives to trigger
another operation which sees the missing replica.  But in the other five cases, there
is evidently a race which is preventing the compensation by resilience for a
premature remove.

_I am at present unsure at what point in the code this needs to be addressed; obviously there
is something which needs to be atomic which currently is not._

**However, the workaround currently would be to scan the target pool immediately upon 
migration move termination.**

**7) How do we account for the rest of resilience activity?** 
 
In the case of noops, resilience does not react before both the Add Cache and Clear Cache messages have arrived, so the 
idempotency of the move is fully seen.

For copy, one would expect resilience to make a compensatory replica for the shared files (123),
since migration move does not make a copy if the file exists on the target, but only deletes
the source.  Hence resilience will see that the count is missing a replica and handle it.

But there are 235 copies, meaning 235 - 123 = 112 extra copy operations.  Where do these come from?

```
Alberts-MacBook-Pro:drain arossi$ for f in `cat copy.txt`; do line=`grep $f shared-copied.txt`; if [ "$line" == '' ] ; then echo $f >> not-shared.txt; fi; done

Alberts-MacBook-Pro:drain arossi$ wc -l not-shared.txt 
     112 not-shared.txt

Alberts-MacBook-Pro:drain arossi$ grep dcatest not-shared-loc.txt | cut -d ' ' -f4 | wc -l
     112
```

These all have three locations. And they are all in the "not-sticky" locations:

```
Alberts-MacBook-Pro:drain arossi$ for f in `cat not-shared.txt`; do line=`grep $f not-sticky.txt`; if [ "$line" == '' ] ; then echo "$f N"; else echo "$f Y"; fi; done | grep Y | wc -l
     112
```

Here is what I surmise is happening in this case:  the AddCacheLocation and ClearCacheLocation messages from the migration move arrive in reverse order; or at least the updates to the namespace are seen by Resilience in reverse order. Seeing the deletion first, which appears to leave only one location (not on the target pool, since these are not shared files), Resilience makes another copy.  Then it sees the new target location, and caches one of the three locations.

**So it is possible to provoke some minor "thrashing".  In this particular run, it occurs for about 10% of the file locations.**

**Comment by alrossi** [comment]
Hmm.  I started to look at the code to see if I could figure out where in Resilience I may need to guarantee atomicity to take care of the lost replication, and it occurred to me that my operation count strategy might be at fault.   

This actually suggests that I may be able to verify this by doing the same thing with the new QoS Engine, which does not use a counting algorithm to determine whether more work needs to be done.   I will do that tomorrow and report back.

**Comment by alrossi** [comment]
I ran an identical test on the new qos engine setup.  I am pleased to report that the results suggest this problem has disappeared with my new queue semantics.

Starting with pools purged of cached files, after the migration move,

```
echo "\s dcatest* rep ls" | ssh -p 24223 admin@fndcatemp2.fnal.gov | grep "si" | grep -v X
Pseudo-terminal will not be allocated because stdin is not a terminal.
    0000EDC5D10AB24E49E0A09B859BF52DF4EA <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}

[fndcatemp2] (local) admin > \sn cacheinfoof 0000EDC5D10AB24E49E0A09B859BF52DF4EA
 dcatest07-3 dcatest07-4 dcatest09-3
```

This means a single premature copy was done, and then the extra copy was cached.

There are no unreplicated files:

```
for f in `cat 03-3-before`; do found=`grep $f verified.txt`; if [ "$found" == '' ] ; then echo "$f N" >> results.txt; else echo "$f Y" >> results.txt; fi; echo "\sn cacheinfoof $f" | ssh -p 24223 admin@fndcatemp2.fnal.gov >> results.txt; done

[arossi@fndcatemp1 ~]$ grep dcatest results.txt | cut -d ' ' -f2 | grep dcatest | wc -l 
975
[arossi@fndcatemp1 ~]$ grep dcatest results.txt | cut -d ' ' -f3 | grep dcatest | wc -l 
975
[arossi@fndcatemp1 ~]$ grep dcatest results.txt | cut -d ' ' -f4 | grep dcatest | wc -l 
1
```

The numbers show that all files that were handled have two locations, and one (the one above) has three.  So all of the others are sticky.

In additions, only the files which were not shared between source and target were handled.

```
[arossi@fndcatemp1 ~]$ grep Y results.txt | wc -l
118

[arossi@fndcatemp1 ~]$ grep N results.txt | wc -l
857

[arossi@fndcatemp1 ~]$  echo "\s qos-verifier verify history" | ssh -p 24223 admin@fndcatemp2.fnal.gov | cut -d ' ' -f3 | cut -c 2-37 > verified.txt
[arossi@fndcatemp1 ~]$ wc -l verified.txt 
118 verified.txt
```

I attribute the superior handling to the fact that the engine no longer tries to keep a count of how many iterations on an operation it thinks it needs to do; I switched to a simple:  do something, recheck, ... exit, entirely determined by the results of the recheck.  Much simpler and more reliable.

How to fix resilience in the meantime?  I think the major flaw lies here in not allowing clear cache location messages to be processed as "full" (that is, iterate as many times as necessary), unlike add cache location messages.  I will change this and retest to see if that make a difference.

**Comment by alrossi** [comment]
Repeating the above experiment with the standalone qos cell.

After migration, there are more cached files than in the distributed case:

```
Alberts-MacBook-Pro:drainqos arossi$ echo "\s dcatest* rep ls" | ssh -p 24223 admin@fndcatemp2.fnal.gov | grep "si" | grep -v X | cut -d ' ' -f5 > cached.txt

Alberts-MacBook-Pro:drainqos arossi$ wc -l cached.txt 
      41 cached.txt
```

However, all of these are files with three locations:

```
for f in `cat cached.txt`; do echo "\sn cacheinfoof $f" | ssh -p 24223 admin@fndcatemp2.fnal.gov >> cached-locations.txt; echo $f >> cached-locations.txt; done

Alberts-MacBook-Pro:drainqos arossi$ grep dcatest cached-locations.txt | cut -d ' ' -f4 | wc -l
      41
```
So timing can affect the number of premature copies done by qos/resilience, but the new queue semantics does seem to have fixed the problem of incomplete replication.

_**What can, however, happen, is that this timing issue interferes with the proper partitioning of copies**_.

For instance, after the migration, one can find instances like

```
0000904F395B4A714C71BA4E3A561003C7E8 N
 dcatest08-3 dcatest08-4

[fndcatemp2] (qos@qosDomain) admin > \s dcatest08-3,dcatest08-4 rep ls 0000904F395B4A714C71BA4E3A561003C7E8
dcatest08-3:
    0000904F395B4A714C71BA4E3A561003C7E8 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest08-4:
    0000904F395B4A714C71BA4E3A561003C7E8 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}
```

Only after a scan is rerun will this become:

```
[fndcatemp2] (qos@qosDomain) admin > \s dcatest07-4,dcatest08-3,dcatest08-4 rep ls 0000904F395B4A714C71BA4E3A561003C7E8
dcatest07-4:
    0000904F395B4A714C71BA4E3A561003C7E8 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest08-3:
    0000904F395B4A714C71BA4E3A561003C7E8 <C-------X--L(0)[0]> 1 si={persistent.dcache-devel-test}
dcatest08-4:
    0000904F395B4A714C71BA4E3A561003C7E8 <C----------L(0)[0]> 1 si={persistent.dcache-devel-test}
```

This does indeed suggest that it is possible for qos/resilience to see the delete in the namespace before it sees the new location, and thus choose another pool on the same host when the storage group (such as this one) specifies not to.

I am trying to figure out why this does not get caught immediately during re-verification.

**Comment by alrossi** [comment]
Here is the task statistics breakdown from before rerunning scans:

```
POOL                         |          TGT       FAILED |        CPSRC        CPTGT        CPBYT |        CACHE      PERSIST       UNSETP |        STAGE        FLUSH
dcatest03-3                  |           62            0 |            0           62           62 |            0            0            0 |            0            0
dcatest03-4                  |            0            0 |          310            0            0 |            0            0            0 |            0            0
dcatest04-3                  |           91            0 |            0           91           91 |            0            0            0 |            0            0
dcatest04-4                  |           95            0 |            0           95           95 |            0            0            0 |            0            0
dcatest07-3                  |           91            0 |            0           91           91 |            0            0            0 |            0            0
dcatest07-4                  |          143            0 |            0          143          143 |            0            0            0 |            0            0
dcatest08-3                  |            0            0 |          326            0            0 |            0            0            0 |            0            0
dcatest08-4                  |           41            0 |           25            0            0 |           41            0            0 |            0            0
dcatest09-3                  |           95            0 |            0           95           95 |            0            0            0 |            0            0
dcatest09-4                  |           84            0 |            0           84           84 |            0            0            0 |            0            0
```

This shows that the copy on dcatest08-3 was originally there, for that pool is used as source, but never as target.  

My reconstruction is as follows:

1. the migration move of the file and the delete occur quickly enough for both to be registered in the namespace by the time the messages hit qos.  
2. the initial triage done to determine if action should be taken does not take into consideration the partitioning logic.
3. qos sees the proper number of copies and so it does not process the file.

Step 2:  the partitioning logic is part of a more involved verification procedure which is done with the operation runs, and that work should not be done on the thread that is responsible for initially processing the message into a potential operation.  The flaw suggests only one solution:  do not do any triage at all, but just translate all such messages into operations, and then let the full verification happen.  Short of that solution, we will just have to compensate for possible inconsistencies from migration move by rescanning the target pool immediately following.

**Comment by alrossi** [comment]
The small fix I though might help in resilience still does not prevent insufficient replication from happening.   

```
Alberts-MacBook-Pro:drainqos arossi$ grep dcatest cached-locations.txt | cut -d ' ' -f2 | grep dcatest | wc -l
    1191
Alberts-MacBook-Pro:drainqos arossi$ grep dcatest cached-locations.txt | cut -d ' ' -f3 | grep dcatest | wc -l
    1191
Alberts-MacBook-Pro:drainqos arossi$ grep dcatest cached-locations.txt | cut -d ' ' -f4 | grep dcatest | wc -l
    1189
```

After another drain, two files have only two replicas, one of which is cached.

**Comment by alrossi** [comment]
I have finished rerunning integration tests over the modified QoS engine (which eliminates triage in all but one case).   Everything looks good.   It also handles the drain correctly (rechecked this).

As for resilience, I don't believe there is an easy way of fixing the little crack we have discovered short of back-porting a rather involved change developed in connection with the new QoS handling.

I would thus recommend as best practice to those working with resilience to scan the target pool(s) after a migration.  This will ensure that the additional replicas are there expeditiously and do not have to await the periodic scan interval to be made.
